cmd/metal-api/internal/service/async-actor.go

```diff
@@ -0,0 +1,167 @@
+package service
+
+import (
+	"errors"
+	"fmt"
+
+	ipamer "github.com/metal-stack/go-ipam"
+	"github.com/metal-stack/metal-api/cmd/metal-api/internal/datastore"
+	"github.com/metal-stack/metal-api/cmd/metal-api/internal/ipam"
+	"github.com/metal-stack/metal-api/cmd/metal-api/internal/metal"
+	"github.com/metal-stack/metal-lib/bus"
+	"go.uber.org/zap"
+)
+
+type asyncActor struct {
+	*zap.Logger
+	ipam.IPAMer
+	*datastore.RethinkStore
+	machineReleaser bus.Func
+	ipReleaser      bus.Func
+}
+
+func newAsyncActor(l *zap.Logger, ep *bus.Endpoints, ds *datastore.RethinkStore, ip ipam.IPAMer) (*asyncActor, error) {
+	actor := &asyncActor{
+		Logger:       l,
+		IPAMer:       ip,
+		RethinkStore: ds,
+	}
+	var err error
+	_, actor.machineReleaser, err = ep.Function("releaseMachineNetworks", actor.releaseMachineNetworks)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create async bus function for machine releasing: %w", err)
+	}
+	_, actor.ipReleaser, err = ep.Function("releaseIP", actor.releaseIP)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create bus function for ip releasing: %w", err)
+	}
+	return actor, nil
+}
+
+func (a *asyncActor) freeMachine(pub bus.Publisher, m *metal.Machine) error {
+	if m.State.Value == metal.LockedState {
+		return fmt.Errorf("machine is locked")
+	}
+
+	err := deleteVRFSwitches(a.RethinkStore, m, a.Logger)
+	if err != nil {
+		return err
+	}
+
+	err = publishDeleteEvent(pub, m, a.Logger)
+	if err != nil {
+		return err
+	}
+
+	// call the releaser async
+	err = a.machineReleaser(m)
+	if err != nil {
+		// log error, but what should we do here? we already called
```

Comment:
Can't we put `deleteVRFSwitches`, `publishDeleteEvent`, `machineUpdate` all into the machineReleaser and make this idempotent? Such that the api tries to accomplish all these things again and again until success? Does this have any bad side effects?

Reply:
the `publishDeleteEvent` must not be retried if anything fails. if in a later (async) step the release of an ip address fails the machine will already be released. so it is possible that the machine is already allocated by another one. so we must not publish the delete event again.

the `deleteVRFSwitches` is another part. i'm not quite deep in the code but i think this has to be done before the machine is deleted.

Reply:
I still do not fully get why we do not put `UpdateMachine` into the reconcile. Did you put it there because you wanted to make a clear statement for the consumer that after this endpoint was called it is guaranteed that this machine has no allocation anymore?  

If the call fails there, the user needs to call `machineFree` again but there are already reconcile loops working on the deletion. I don't know if it is good to motivate the user calling this endpoint over and over again. Also, wouldn't it be a little more robust to put `UpdateMachine` in the reconcile to make sure that networks were cleaned up properly before allowing anyone else to allocate this machine again? I think, at most providers, when you call the deletion of a machine, you will continue to see it until it was fully removed by the backend.

And if we put `UpdateMachine` in the loop, we also need to put `deleteVRFSwitches` and `publishDeleteEvent` in the loop because otherwise the machine could already have booted into the old OS (hanging in `Phoned Home`). So we need to call both functions right before `UpdateMachine`. `UpdateMachine` needs to come last because after that another user can allocate the machine again, which should be the last step to finish off deletion.

Reply:
well, imho the `metal-api` should not care about callers. if the caller of the api has a reconcile loop this is a valid strategy; it would also be a strategy to handle errors in another way. 

and the caller has to handle errors. the deletion of a machine **can** fail: we check for a locked machine (perhaps we will have more validations in the future) or we can have technical problems communicating with the DB or NSQ. 

we have to make a balance between the promise to the caller that the deletion really works and some errors which can happen. doin'g everything async will always return an OK to the caller in any case (except when we have a communication error with NSQ).

i think when the `machineFree` returns, the machine should be unallocated. so the caller of this function can be sure that the machine is destroyed/unallocated. the dependent ressources like IP adresses will be cleaned up async.  this guarantees some sort of consistency: after a successfull deletion of a machine the caller can immediatly query the same machine and the api should return an unallocated machine.

the current code does this in a sync way:
  1. remove the machine from the network
  2. send a deletion event to metal-core (triggers a reboot)
  3. updates the allocation state

now the machine should be free; and it also forks an async cleaup-job for all the ressources which are still allocated by the machine.
i'm thinking about switching the order of 2. and 3. because metal-core triggers a reboot, and the machine then queries if it is allocated; this should return nil; but the current code can not guarantee this.

doin'g the unallocation synchron makes sure that the caller can be sure that the machine is free after a successfull call. and i think this is the intention of the caller.

but i see your point too :smiley: 

doin'g anything in a async block would work too. but than we change the semantic of our api. this would be the first endpoint which can be called successfully but if you query the api direct after the `free` it still returns the allocation for this machine. so a fully async implementation adds some form of inconsistency to our api. i'm not sure if we want this (here).

in this case the caller would need some form of a promise (a callback) which will be called after the machine is really free. or the caller polls until the machine is really unallocated. not every client of this api is a machine reconciler which does not care about the machine any more ...

Reply:
Cool, thanks for your detailed explanation. It now makes more sense to me that this is the right approach for the time being.

If at some point in the future we still observe a need for more robustness, I think that we could discuss again if it wouldn't be better to put as much into the reconciliation as possible (which probably feels a bit more like dealing with "kubernetes"). I can imagine that more tasks in the reconciliation can give us more control over our workflows and the final state can be reached more reliably without getting into states of inconsistent data.